graph/options/options.go

```diff
@@ -47,7 +47,7 @@ type VendorOptions struct {
 
```

Comment:
@israel-hdez , I haven't looked too deeply at the metrics side of things, I wanted to start with the graph code.  My initial comment is that there may be too much logic being deferred to the lower levels of the code.  I may be wrong but I think we could handle most of it right here in the options pre-processing.    We know here which namespaces are requested, as well as the queryTime (i.e. endTime) and the duration.  The difference now is that we may have a different duration for each namespace, right?   Instead of AccessibleNamespaces having the side-effect of carrying namespace creation time, maybe keep that info local to this file and instead change Options.Namespaces to a map[string]time.Duration (or maybe it could be a map to new struct like map[string]NamespaceInfo). Then populate the map by calculating a valid duration per requested namespace (based on the requested duration, queryTime and namespace creation time).  You can then remove Options.Duration.   If you do this I think the downstream impact would be almost negligible.  The only difference would be replacing accesses to options.Duration with something like options.Namespaces[<namespace>].Duration.

What do you think?

Reply:
@jshaughn Yes, indeed with your suggestion makes sense. It will minimize changes in the inner levels of the code.

I was trying to see the code with a 3-tier architecture, since there already are some "business" and "models" packages. In my point of view, the options parsing step is still part of the "presentation" tier and, probably, that was the reason I didn't do it here (I was trying to not mix business with presentation logic).

But... here in the options the namespaces are already being filtered out if they are not accessible which I see also as part of "business logic". So, I think it doesn't hurt if I move the logic to determine per-namespace duration to the parsing of the options. I'll do that.

Reply:
@jshaughn Changed!... I'll think how to change the metrics part.

Reply:
I understand your thinking about the tiers.  And it can be a little confusing because the graph code and metrics code grew somewhat independently.  As for the graph code, I guess I don't really consider any of the server-side code as part of the presentation layer.   I think of it mainly as a separation of concerns.  "Options" is concerned with describing the graph to be generated, it is the nexus between the user's request and the graph code.  It can immediately fail given an impossible request, or it can massage the request as needed.  The graph code is responsible for generating the logical graph topology and should not be concerned with the graph definition.  The appenders are responsible for optional decoration of the graph, and the vendor code (we only have the one cytoscape vendor now, but we used to have vizceral) translates the graph into a particular format for consumption by the caller.